doc/ip/next_hop_group_hld.md

```diff
@@ -0,0 +1,148 @@
+
+
+# About this Manual
+This document provides general information about an enhancement to the internal APP_DB routing table to split next hop information out into its own table.
+
+Throughout this document a next hop group may refer to either a single next hop or a group of several next hops.
+
+# 1 Introduction
+Currently the route table within APP_DB contains all next hop information for a route embedded in that route's entry in the table.
+
+At high scale (particularly when handling millions of routes all routed over multiple next hops) this is inefficient both in terms of performance and occupancy.  A given route entry needs to be built up including all its next hop information, which gets sent to and stored in APP_DB.  Orchagent then needs to receive, parse and handle all the information in that route entry every time.
+
+A more efficient system would involve managing the next hop groups in use by the route table separately, and simply have the route table specify a reference to which next hop group to use.  Since at scale many routes will use the same next hop groups, this requires much smaller occupancy per route, and so more efficient building, transmission and parsing of per-route information.
+
+# 2 Requirement Overview
+## 2.1 Functional Requirements
+
+- Improve efficiency of storage of routing information in APP_DB.
+- Reduce time to reprogram routes when next hop information changes.
+
+## 2.2 Configuration and Management Requirements
+- The existing 'show ip route' and 'show ipv6 route' commands must be supported when using the new next hop group table.
+
+## 2.3 Scalability Requirements
+- Unchanged in requirements, but will improve scalability in terms of storage and programming time.
+
+## 2.4 Warm Boot Requirements
+- Unchanged - the new next hop group table must be compatible with existing warm boot requirements.
+
+# 3 Design
+## 3.1 Overview
+This design directly changes APP_DB and orchagent, along with minor changes to support the existing CLI when using the new APP_DB table.
+
+## 3.2 DB Changes
+### 3.2.1 APP DB
+Currently the ROUTE_TABLE in APP_DB includes all next hop information for the route, as follows:
+```
+### ROUTE_TABLE
+    ;Stores a list of routes
+    ;Status: Mandatory
+    key           = ROUTE_TABLE:prefix
+    nexthop       = *prefix, ;IP addresses separated “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated by “,” (zero indicates no interface)
+    blackhole     = BIT ; Set to 1 if this route is a blackhole (or null0)
+ ```
+
+This design adds a new NEXT_HOP_GROUP_TABLE, to store next hop group information to be used by one or more routes.
+```
+### NEXT_HOP_GROUP_TABLE
+    ;Stores a list of groups of one or more next hops
+    ;Status: Mandatory
+    key           = NEXT_HOP_GROUP_TABLE:string ; arbitrary identifier for the next hop group
+    nexthop       = *prefix,           ; IP addresses separated “,” (empty indicates no gateway)
```

Comment:
>nexthop [](start = 4, length = 7)

How one next hop group is shared by multiple routes?

Reply:
Let me clarify my questions:
1. Will you reuse the same entry in NEXT_HOP_GROUP_TABLE for multiple routes?
2. If yes, you may use nexthop as the 'key', and there is no need to use another arbitrary identifier

Reply:
Yes to 1. For 2, the point of this enhancement is to not have the route table store the full set of next hops, and instead use a much smaller identifier. Using the nexthop field as the key to the table would reverse this. Also, it's possible under this system to change the set of next hops within a group while keeping the identifier the same, which wouldn't be possible if the next hops were the identifier.